validator/client/src/main/java/tech/pegasys/teku/validator/client/loader/MultithreadedValidatorLoader.java

```diff
@@ -0,0 +1,81 @@
+ */
+
+package tech.pegasys.teku.validator.client.loader;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.infrastructure.logging.StatusLogger;
+import tech.pegasys.teku.validator.api.GraffitiProvider;
+import tech.pegasys.teku.validator.client.Validator;
+import tech.pegasys.teku.validator.client.loader.ValidatorSource.ValidatorProvider;
+
+public class MultithreadedValidatorLoader {
+
+  public static Map<BLSPublicKey, Validator> loadValidators(
+      final Map<BLSPublicKey, ValidatorProvider> providers,
+      final GraffitiProvider graffitiProvider) {
+    final int totalValidatorCount = providers.size();
+    StatusLogger.STATUS_LOG.loadingValidators(totalValidatorCount);
+
+    final ExecutorService executorService =
+        Executors.newFixedThreadPool(Math.min(4, Runtime.getRuntime().availableProcessors()));
+    try {
+      final AtomicInteger numberOfLoadedKeys = new AtomicInteger(0);
+      final List<Future<Validator>> futures =
+          providers.values().stream()
+              .map(
+                  provider ->
+                      executorService.submit(
+                          () -> {
+                            final Validator validator =
+                                new Validator(
+                                    provider.getPublicKey(),
+                                    provider.createSigner(),
+                                    graffitiProvider);
+                            int loadedValidatorCount = numberOfLoadedKeys.incrementAndGet();
+                            if (loadedValidatorCount % 10 == 0) {
+                              StatusLogger.STATUS_LOG.atLoadedValidatorNumber(
+                                  loadedValidatorCount, totalValidatorCount);
+                            }
+                            return validator;
+                          }))
+              .collect(toList());
+
+      final Map<BLSPublicKey, Validator> validators = new HashMap<>();
+      for (Future<Validator> future : futures) {
+        final Validator validator = future.get();
+        validators.put(validator.getPublicKey(), validator);
+      }
+      return validators;
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Interrupted while attempting to load validator key files", e);
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException) {
+        throw (RuntimeException) e.getCause();
+      }
+      throw new RuntimeException("Unable to load validator key files", e);
+    } finally {
+      executorService.shutdownNow();
+    }
```

Comment:
(nit) May be `Stream.parallel()` would yield a nicer code? 
```suggestion
    return providers.values().stream()
        .parallel()
        .map(
            provider ->
                new Validator(provider.getPublicKey(), provider.createSigner(), graffitiProvider))
        .peek(
            __ -> {
              int loadedValidatorCount = numberOfLoadedKeys.incrementAndGet();
              if (loadedValidatorCount % 10 == 0) {
                StatusLogger.STATUS_LOG.atLoadedValidatorNumber(
                    loadedValidatorCount, totalValidatorCount);
              }
            })
        .collect(toMap(Validator::getPublicKey, Function.identity()));
```

Reply:
BTW `stream().parallel()` could be reduced to `parallelStream()`

Reply:
This code just moved - it deliberately avoids just using `parallelStream()` to limit the number of concurrent operations.  Decrypting scrypt keystores requires a significant amount of memory so if you have a lot of CPUs `parallelStream()` decrypts too many in parallel and throws an `OutOfMemoryError`.  It would be nice to adapt the limit based on how much heap space is available but we've never gotten around to it.  Could also use a custom ForkJoinThreadPool but it doesn't come out much nicer and is way harder to understand.

Reply:
Oh, got it :/

Reply:
May be it's worth inlining a comment why we are limiting threads, because the reason is non-trivial and could be considered later

Reply:
Done.  It was more confusing given it had moved to its own class as well I think.